src/backend/distributed/planner/multi_explain.c

```diff
@@ -199,6 +201,20 @@ CitusExplainScan(CustomScanState *node, List *ancestors, struct ExplainState *es
 		return;
 	}
 
+	if (SkipConstraintValidation || (AlterTableInProgress() && IsCitusInternalBackend()))
+	{
+		/*
+		 * We do not want to explain ALTER TABLE ... ADD FOREIGN KEY .. queries in
+		 * in the workers. They trigger a SELECT query on the citus tables as part of the
```

Comment:
I was actually going to propose 

   if (AlterTableInProgress() && SkipConstraintValidation)

instead. 

 My reasoning goes like this:

    if (AlterTableInProgress() && SkipConstraintValidation) 
           -  will never explain the validation steps for shell tables since SkipConstraintValidation is always set by citus for them.
           - will explain the validation on shards if the validation is not skipped explicitly by the client.

I believe this condition best captures the intention assuming the validation on shards do not cause distributed transaction unlike validation on citus (shell) tables.

Reply:
> - will never explain the validation steps for shell tables since SkipConstraintValidation is always set by citus for them.

Is this accurate? Do we always send it to workers AND set it on the coordinator? Maybe true, but as far as I can see we don't send it to workers for all operations, we only send it for certain code-paths in the rebalancer ?

Reply:
As for my debugging:

    On a worker process: 
         SkipConstraintValidation is always set on implicitly by the EnableSkippingConstraintValidation function for the citus/distributed/shell tables. Hence it is always true for 

 ` SELECT fk.\"col_1\" FROM ONLY \"public\".\"target_table\" fk LEFT OUTER JOIN ONLY \"public\".\"test_ref_table\" pk ON ( pk.\"key\" OPERATOR(pg_catalog.=) fk.\"col_1\") WHERE pk.\"key\" IS NULL AND (fk.\"col_1\" IS NOT 
    `
     queries.

         SkipConstraintValidation is not overriden for the shard tables. Hence it could be true or false. Those queries are not distributed and do not get explained by CitusExplainScan (I was wrong mentioning enabling those in CitusExplainScan above)

     In CitusExplainScan, we are trying to mimic AlterTableConstraintCheck which  seems to be the function for deciding if a statement is a constraint check or not. 

   Now my question is :

   Since we cannot exactly express AlterTableConstraintCheck  within CitusExplainScan, can we choose to skip all ALTER TABLE statements with the following condition check?
 
  if (AlterTableInProgress())